homeworks/MarlQ/1-halfedge/src/geometry/halfedge.ts

```diff
@@ -0,0 +1,260 @@
+    this.color = new Vector(0, 128, 255, 1);
+    this.wireframe = new Vector(125, 125, 125, 1);
+
+    // load .obj file
+    const indices: number[] = [];
+    const positions: Vector[] = [];
+    const lines = data.split('\n');
+    for (let line of lines) {
+      line = line.trim();
+      const tokens = line.split(' ');
+      switch (tokens[0].trim()) {
+        case 'v':
+          positions.push(
+            new Vector(
+              parseFloat(tokens[1]),
+              parseFloat(tokens[2]),
+              parseFloat(tokens[3]),
+              1
+            )
+          );
+          break;
+        case 'f':
+          // only load indices of vertices
+          for (let i = 1; i < tokens.length; i++) {
+            const vv = tokens[i].split('/');
+            indices.push(parseInt(vv[0]) - 1);
+          }
+          break;
+      }
+    }
+
+    this.verts = []; // an array of Vertex object
+    this.edges = []; // an array of Edge object
+    this.faces = []; // an array of Face object
+    this.halfedges = []; // an array of Halfedge object
+    this.buildMesh(indices, positions);
+  }
+
+  /**
+   * buildMesh builds half-edge based connectivity for the given vertex index buffer
+   * and vertex position buffer.
+   *
+   * @param indices is the vertex index buffer that contains all vertex indices.
+   * @param positions is the vertex buffer that contains all vertex positions.
+   */
+  buildMesh(indices: number[], positions: Vector[]) {
+    // TODO: build the halfedge connectivity.
+    //
+    // We can assume the input mesh is a manifold mesh.
+    
+    // DEBUG
+    // let duplicates = 0;
+    // indices = indices.slice(0, 120); 
+
+    positions.forEach((p, i) => {
+      const vert = new Vertex(p);
+      vert.idx = i;
+      this.verts.push(vert);
+    });
+
+    for(let index = 2; index < indices.length; index+=3) {
+      const face = new Face();
+      face.idx = Math.floor(index/3);
+      this.faces.push(face);
+
+      const halfedges: Halfedge[] = [];
+
+      for(let edge_count = 0; edge_count < 3; edge_count++) {
+        let edge: Edge;
+
+        let halfedge: Halfedge;
+
+        let vertex_from = this.verts[indices[index - 2 + edge_count]];
+        let vertex_to = this.verts[indices[index - 2 + ((edge_count+1) % 3)]];
+
+        halfedge = new Halfedge();
+        halfedge.onBoundary = true;
+
+        
+        let existing_twin = this.halfedges.find( aV => aV.vert?.idx === vertex_to.idx && aV.next?.vert?.idx === vertex_from.idx);
+        if(existing_twin) { // Twin halfedge exists
+          //duplicates++ // DEBUG
+
+          halfedge.onBoundary = false;
+          existing_twin.onBoundary = false;
+
+          halfedge.twin = existing_twin;
+          existing_twin.twin = halfedge;
+
+          if(halfedge.twin.edge) edge = halfedge.twin.edge;
+          else {
+            console.log("ERROR: NO EDGE ON TWIN"); // Should never occur (thanks Typescript)
+            edge = new Edge();
+          }
```

Comment:
If this does not occur, we could simply document it and remove the else case.

Reply:
Typescript wouldn't let me remove the else.